### PR TITLE
Implement support for caddy placeholders in replacement

### PR DESCRIPTION
**Pull request by fanste**

### CODE REVIEW

File: filter.go [hunk 1]
```diff
@@ -22,6 +23,8 @@ func init() {
 	httpcaddyfile.RegisterHandlerDirective("filter", parseCaddyfile)
 }
 
+var paramReplacementPattern = regexp.MustCompile("\\{[a-zA-Z0-9_\\-.]+}")
```
Comment (htfy96) [inline filter.go]: Can you make the `+` non-greedy by changing it to `+?`? I would expect it to be faster if we use non-greedy matches

File: filter.go [hunk 2]
```diff
@@ -173,10 +212,13 @@ func (m Middleware) ServeHTTP(w http.ResponseWriter, r *http.Request, next caddy
 	csr.FlushHeaders()
 	if m.compiledContentTypeRegex.MatchString(csr.Recorder().Result().Header.Get("Content-Type")) {
 		buf := new(bytes.Buffer)
+		repl := r.Context().Value(caddy.ReplacerCtxKey).(*caddy.Replacer)
 		if _, err := buf.ReadFrom(csr.Recorder().Result().Body); err != nil {
 			return fmt.Errorf("failed to read from response body: %w", err)
 		}
-		replaced := m.compiledSearchRegex.ReplaceAll(buf.Bytes(), []byte(m.Replacement))
+		replaced := m.compiledSearchRegex.ReplaceAllFunc(buf.Bytes(), func(input []byte) []byte {
```
Comment (htfy96) [inline filter.go:227]: Perhaps I missed something, but is it possible to generate the replacement string ahead of time?

For example:

```go
// When m.Replacement is "abc{system.os}def"
realReplacement := paramReplacementPattern.ReplaceAllFunc(replacement, func(input2 []byte) []byte {
		return m.paramReplacer(input2, groups, repl)
	})
// realReplacement is "abclinuxdef"
replaced := m.compiledSearchRegex.ReplaceAll(buf.Bytes(), realReplacement)
```

File: filter.go [hunk 3]
```diff
@@ -160,6 +163,42 @@ func (csr *CappedSizeRecorder) WriteHeader(statusCode int) {
...
+		return []byte{}
+	}
+	groups := pattern.FindSubmatch(input)
+	replacement := paramReplacementPattern.ReplaceAllFunc(rawReplacement, func(input2 []byte) []byte {
+		return m.paramReplacer(input2, groups, repl)
+	})
+	return replacement
+}
+
+func (m Middleware) paramReplacer(input []byte, groups [][]byte, repl *caddy.Replacer) []byte {
```
Comment (htfy96) [inline filter.go:188]: Add some document here saying it supports both `{caddy_replacer}` and capture group replacement like `{1}`

**Review by htfy96 — CHANGES_REQUESTED** [review]
Thanks for contribution. Can you add some test cases in `test/` directory covering both `{caddy_replacement}` and `{1}` use cases?

File: filter.go (hunk 1 above)
Comment (fanste) [reply filter.go]: According to some online blog posts, that should not change anything, as the group only allows a limited set of characters. But I will change it, if you feel more comfortable with it.

File: filter.go (hunk 2 above)
Comment (fanste) [reply filter.go:227]: That makes no sense, as the caddy placeholders also contain request-specific, reverse-proxy-specific, ... placeholders. If we generate the replacement string ahead of time (I think you mean inside `Provision`?), than we won't be able to use those placeholders. And thats the usecase that I need. The code is btw similar to the one of caddy v1 filter module.

File: filter.go (hunk 3 above)
Comment (fanste) [reply filter.go:188]: Done, also added a comment to the README

**Comment by fanste** [comment]
> 
> 
> Thanks for contribution. Can you add some test cases in `test/` directory covering both `{caddy_replacement}` and `{1}` use cases?

I have no linux machine or emulator to test the `run-test` script, but I tried to verify the test implementation using windows tools

File: filter.go (hunk 2 above)
Comment (htfy96) [reply filter.go:227]: I see. Thanks for explaining this

**Review by htfy96 — APPROVED** [review]
lgtm. Thanks for spending time improving this!

**Pull request merged**